### DISCUSSION TITLE
What is the best way to make a Body that can be constrained not have collisions?

### DISCUSSION THREAD

**Discussion by DennisSmolek** [jrouwe/JoltPhysics.js#132] [Q&A]
I’m working on camera rigs that use constraints to attach to objects like player characters and I don’t want them to be collidable but they need to go through the physics system to work the constraints correctly.

What I’ve done is create a new layer for them and used an ObjectFilter and disable collisions with every other layer.

This seems correct, but also kinda heavy.

I know if I put isSensor on the body then it won’t collide, but I’m not sure you can have dynamic sensors..

I also know you can use collision groups so the bodies don’t hit each other, however I don’t want them to hit ANYTHING. 

is there a better way to make a body be ignored by everything or is the layers the best method?

**Comment by PhoenixIllusion** [comment]
Another option would be to just put it on it's own layer all together, and it will never interact with anything else, although anything else you put in that layer in the future may collide with it. Possibly an interesting bug that could surface in the future.

Regarding sensors, they should work. Sensors have some special behavior if you want them to "sense" things and have collision triggers, regarding static and dynamic, but just having them reject collisions should not require any custom logic. The source code states they can be static, dynamic, or kinematic, so they should work in your case, and you have not stated you even need the sensing part (collision detection)

Most of the special cases are listed in the source code around Sensors:
https://github.com/jrouwe/JoltPhysics/blob/0191c72d5f6a66217df02c0f6ad5d1c63134b4d4/Jolt/Physics/Body/Body.h#L72


Most camera setups that are tied to physics want collision, since one of the main reasons cameras enable physics is to avoid them passing through walls and other objects and messing with the view. Even in those cases, some of them make them live outside the physics system and just use shape-cast and ray-project to ensure that their N+1 position is not intersecting the world, sometime checking the graphics world and not physics world.

Are you using the constraint solver to use springs and non-user controlled motion to the rig, possibly mass and inertia to the rig motion? I could see having the target object move and then having the rig yanked could be an interesting effect.

If you are not using any particular physics on the camera, then another option is to use graphics-engine binding instead of physics-engine binding, such as via an existing graphics engine, and making the root of the camera-rig bound to the character, or by using the raw math of a character's physics position/rotation and then matrix multiplying to to reposition the rig's root.

**Comment by DennisSmolek** [comment]
Right now I have the bodies in their own layer and I put the object filter to disable collision on every layer including its own. This seems to work.

it’s hard to explain and I might not use the physics system for every part yet I’m not sure.

but yeah exactly, the rig uses motion and springs/inertia to orient and keep clear.

 it uses the physics system and constraints to position a ball end correctly.

this is similar to the “camera boom” in other engines.

I could write most of this in ThreeJS but much of the constraint system solves things more elegantly and I raycast quite a bit too.

I’ll post a demo hopefully today or tomorrow
